ddl/column.go

```diff
@@ -1436,3 +1444,61 @@ func indexInfosToIDList(idxInfos []*model.IndexInfo) []int64 {
 	}
 	return ids
 }
+
+func genChangingColumnUniqueName(tblInfo *model.TableInfo, oldCol *model.ColumnInfo) string {
+	newColumnName := fmt.Sprintf("%s%s", changingColumnPrefix, oldCol.Name.O)
+	newColumnLowerName := strings.ToLower(newColumnName)
+	// Check whether the new column name is used.
+	columnNameMap := make(map[string]bool, len(tblInfo.Columns))
+	for _, col := range tblInfo.Columns {
+		columnNameMap[col.Name.L] = true
+	}
+	suffix := 0
+	for columnNameMap[newColumnLowerName] {
+		suffix++
+		newColumnLowerName = fmt.Sprintf("%s_%d", strings.ToLower(newColumnName), suffix)
+	}
+	if suffix != 0 {
+		newColumnName = fmt.Sprintf("%s_%d", newColumnName, suffix)
+	}
+	return newColumnName
+}
+
+func genChangingIndexUniqueName(tblInfo *model.TableInfo, idxInfo *model.IndexInfo) string {
+	newIndexName := fmt.Sprintf("%s%s", changingIndexPrefix, idxInfo.Name.O)
+	newIndexLowerName := strings.ToLower(newIndexName)
+	// Check whether the new index name is used.
+	indexNameMap := make(map[string]bool, len(tblInfo.Indices))
+	for _, idx := range tblInfo.Indices {
+		indexNameMap[idx.Name.L] = true
+	}
+	suffix := 0
+	for indexNameMap[newIndexLowerName] {
+		suffix++
+		newIndexLowerName = fmt.Sprintf("%s_%d", strings.ToLower(newIndexName), suffix)
+	}
+	if suffix != 0 {
+		newIndexName = fmt.Sprintf("%s_%d", newIndexName, suffix)
+	}
+	return newIndexName
+}
+
+func getChangingColumnOriginName(changingCol *model.IndexColumn) string {
+	colName := strings.ToLower(strings.TrimPrefix(changingCol.Name.O, changingColumnPrefix))
+	// Since the unique colName may contain the suffix number (columnName_num), better trim the suffix.
+	var pos int
+	if pos = strings.LastIndex(colName, "_"); pos == -1 {
+		return colName
+	}
+	return colName[:pos]
+}
+
+func getChangingIndexOriginName(changingIdx *model.IndexInfo) string {
+	idxName := strings.TrimPrefix(changingIdx.Name.O, changingIndexPrefix)
+	// Since the unique idxName may contain the suffix number (indexName_num), better trim the suffix.
+	var pos int
+	if pos = strings.LastIndex(idxName, "_"); pos == -1 {
```

Comment:
Got a problem here:

```sql
>create table if not exists t(c2_0 bigint, unique index idx(c2_0));
>alter table test.t change column c2_0 c2_1 tinyint;
(1105, 'runtime error: index out of range [1] with length 1')
```

Reply:
Got it, should unify the suffix start from 0, so the trim suffix is always necessary. Addressed